### PR TITLE
Remove Edi boundary for async

### PR DESCRIPTION
**Pull request by benaadams**
Switch off EDI boundary for async

Resolves https://github.com/dotnet/coreclr/issues/15779

### CODE REVIEW

**Comment by benaadams** [comment]
@jkotas PTAL

File: src/mscorlib/src/System/Diagnostics/Stacktrace.cs [hunk 1]
```diff
@@ -162,7 +155,7 @@ void IDisposable.Dispose()
             }
         }
 
-        public virtual MethodBase GetMethodBase(int i)
```
Comment (benaadams) [inline src/mscorlib/src/System/Diagnostics/Stacktrace.cs]: No point in virtuals on non-inherited internal class

**Comment by benaadams** [comment]
It use the same method to detect async as the open PR https://github.com/dotnet/coreclr/pull/14655 which seems to be approved approach by C# team https://github.com/dotnet/coreclr/pull/14655#issuecomment-346192988?

**Comment by MattGal** [comment]
@dotnet-bot test cross please

**Comment by benaadams** [comment]
OSX Failure `02:35:06 fatal error: error in backend: IO failure on output stream.`

**Comment by benaadams** [comment]
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test

File: src/mscorlib/src/System/Diagnostics/Stackframe.cs [hunk 2]
```diff
@@ -2,25 +2,19 @@
...
 
-
 using System.Text;
-using System;
-using System.IO;
 using System.Reflection;
 
 namespace System.Diagnostics
 {
-    // There is no good reason for the methods of this class to be virtual.
```
Comment (noahfalk) [inline src/mscorlib/src/System/Diagnostics/Stackframe.cs]: I think we should split the cleanup off to its own PR to make sure it doesn't get squashed together with the other changes during checkin. It helps to avoid having a few important changes hidden amongst numerous refactorings.

File: src/mscorlib/src/System/Diagnostics/Stacktrace.cs [hunk 3]
```diff
@@ -592,12 +575,14 @@ internal String ToString(TraceFormat traceFormat)
...
                             sb.AppendFormat(CultureInfo.InvariantCulture, inFileLineNum, fileName, sf.GetFileLineNumber());
                         }
-                    }
 
-                    if (sf.GetIsLastFrameFromForeignExceptionStackTrace())
-                    {
-                        sb.Append(Environment.NewLine);
-                        sb.Append(SR.Exception_EndStackTraceFromPreviousThrow);
+                        if (sf.GetIsLastFrameFromForeignExceptionStackTrace() && 
+                            !(declaringType?.Assembly.FullName.StartsWith(CoreLib.Name, StringComparison.Ordinal) ?? false) && // Skip EDI boundary for mscorlib
```
Comment (noahfalk) [inline src/mscorlib/src/System/Diagnostics/Stacktrace.cs]: Doing a direct comparison on the assembly object seems likely to be faster and more robust. Something like
static Assembly CorelibAssembly = typeof(object).Assembly?

File: src/mscorlib/src/System/Diagnostics/Stacktrace.cs [hunk 4]
```diff
@@ -498,12 +474,19 @@ internal String ToString(TraceFormat traceFormat)
...
 
-                    Type t = mb.DeclaringType;
+                    Type declaringType = mb.DeclaringType;
+                    bool isAsync = false;
                     // if there is a type (non global method) print it
-                    if (t != null)
+                    if (declaringType != null && declaringType.IsDefined(typeof(CompilerGeneratedAttribute)) &&
+                        (typeof(IAsyncStateMachine).IsAssignableFrom(declaringType)))
+                    {
+                        isAsync = true;
```
Comment (noahfalk) [inline src/mscorlib/src/System/Diagnostics/Stacktrace.cs]: Could you add some perf numbers? I had a vague memory you already benchmarked and it was barely noticeable but just want to confirm. Some customers use stack traces on their hot path and get very sad when we slow them down (https://github.com/Microsoft/dotnet/issues/529)

In this case isAsync only needs to be calculated once we determine IsLastFrameFromForeignException is true, but I know if your other use case you need to check it for every frame anyways.

File: src/mscorlib/src/System/Diagnostics/Stackframe.cs (hunk 2 above)
Comment (benaadams) [reply src/mscorlib/src/System/Diagnostics/Stackframe.cs]: Simplified just to this change

File: src/mscorlib/src/System/Diagnostics/Stacktrace.cs (hunk 3 above)
Comment (benaadams) [reply src/mscorlib/src/System/Diagnostics/Stacktrace.cs]: Done

File: src/mscorlib/src/System/Diagnostics/Stacktrace.cs (hunk 4 above)
Comment (benaadams) [reply src/mscorlib/src/System/Diagnostics/Stacktrace.cs]: Hmm... I tested against throwing the exception and it was insignificant  - but I didn't look at just creating a StackTrace - however this is in the `.ToString` path so wouldn't impact the linked example. 

Never the less, will check

File: src/mscorlib/src/System/Diagnostics/Stacktrace.cs [hunk 5]
```diff
@@ -198,6 +198,8 @@ public virtual bool IsLastFrameFromForeignExceptionStackTrace(int i)
     // There is no good reason for the methods of this class to be virtual.  
     public class StackTrace
     {
+        private static Assembly s_corelibAssembly = typeof(object).Assembly;
```
Comment (stephentoub) [inline src/mscorlib/src/System/Diagnostics/Stacktrace.cs]: readonly?

File: src/mscorlib/src/System/Diagnostics/Stacktrace.cs [hunk 6]
```diff
@@ -498,12 +500,19 @@ internal String ToString(TraceFormat traceFormat)
...
 
-                    Type t = mb.DeclaringType;
+                    Type declaringType = mb.DeclaringType;
+                    bool isAsync = false;
+                    if (declaringType != null && 
+                        declaringType.IsDefined(typeof(CompilerGeneratedAttribute)) &&
+                        typeof(IAsyncStateMachine).IsAssignableFrom(declaringType))
+                    {
+                        isAsync = true;
+                    }
```
Comment (stephentoub) [inline src/mscorlib/src/System/Diagnostics/Stacktrace.cs]:
```
bool isAsync = declaringType != null && ...;
```
?

File: src/mscorlib/src/System/Diagnostics/Stacktrace.cs [hunk 7]
```diff
@@ -594,7 +603,9 @@ internal String ToString(TraceFormat traceFormat)
                         }
                     }
 
-                    if (sf.GetIsLastFrameFromForeignExceptionStackTrace())
+                    if (sf.GetIsLastFrameFromForeignExceptionStackTrace() &&
+                        declaringType?.Assembly != s_corelibAssembly && // Skip EDI boundary for mscorlib
```
Comment (stephentoub) [inline src/mscorlib/src/System/Diagnostics/Stacktrace.cs]: Why is this disabled for all of corelib?

Comment (benaadams) [reply src/mscorlib/src/System/Diagnostics/Stacktrace.cs]: 😞 

Will have to narrow further

Comment (benaadams) [reply src/mscorlib/src/System/Diagnostics/Stacktrace.cs]: Perhaps attribute on method is way forward?

Comment (noahfalk) [reply src/mscorlib/src/System/Diagnostics/Stacktrace.cs]: attribute seems reasonable, and matches the pattern we used for StackTraceHidden handling

Comment (benaadams) [reply src/mscorlib/src/System/Diagnostics/Stacktrace.cs]: Just switched it off for async; can address other areas if/when they arise

**Comment by benaadams** [comment]
Changed to just switched it off for async; can address other areas if/when they arise

**Comment by benaadams** [comment]
Tizen
```
FAILED   - JIT/Regression/VS-ia64-JIT/V1.2-M01/b12263/b12263/b12263.sh
 BEGIN EXECUTION
 /home/coreclr/bin/tests/Windows_NT.x64.Checked/Tests/coreoverlay/corerun b12263.exe
 qemu: Unsupported syscall: 389

qemu: uncaught target signal 11 (Segmentation fault) - core dumped
 ./b12263.sh: line 243:  2459 Segmentation fault      (core dumped) $_DebuggerFullPath "$CORE_ROOT/corerun" $ExePath $CLRTestExecutionArguments
 Expected: 100
 Actual: 139
 END EXECUTION - FAILED
```

**Comment by benaadams** [comment]
Ubuntu arm Cross Debug Innerloop Build 
```
FAILED   - JIT/Regression/CLR-x86-JIT/V1-M09.5-PDC/b32560/b32560/b32560.sh
  BEGIN EXECUTION
  /home/coreclr/bin/tests/Windows_NT.x64.Debug/Tests/coreoverlay/corerun b32560.exe
  ./b32560.sh: line 243:  1837 Segmentation fault      (core dumped) 
    $_DebuggerFullPath "$CORE_ROOT/corerun" $ExePath $CLRTestExecutionArguments
  Expected: 100
  Actual: 139
  END EXECUTION - FAILED
```

**Comment by benaadams** [comment]
@dotnet-bot Tizen armel Cross Checked Innerloop Build and Test
@dotnet-bot Ubuntu arm Cross Debug Innerloop Build

**Comment by benaadams** [comment]
@dotnet-bot test Tizen armel Cross Checked Innerloop Build and Test
@dotnet-bot test Ubuntu arm Cross Debug Innerloop Build

**Comment by noahfalk** [comment]
I'm happy if Stephen is. Thanks!

**Comment by stephentoub** [comment]
@benaadams, can you share a before/after with this change?

**Comment by benaadams** [comment]
Pre
```
System.Exception: Exception of type 'System.Exception' was thrown.
   at asyncexceptions.Program.<RecursiveAsync>d__2.MoveNext() in C:\Work\Things\asyncexceptions\Program.cs:line 32
--- End of stack trace from previous location where exception was thrown ---
   at asyncexceptions.Program.<RecursiveAsync>d__2.MoveNext() in C:\Work\Things\asyncexceptions\Program.cs:line 32
--- End of stack trace from previous location where exception was thrown ---
   at asyncexceptions.Program.<RecursiveAsync>d__2.MoveNext() in C:\Work\Things\asyncexceptions\Program.cs:line 32
--- End of stack trace from previous location where exception was thrown ---
   at asyncexceptions.Program.<RecursiveAsync>d__2.MoveNext() in C:\Work\Things\asyncexceptions\Program.cs:line 32
--- End of stack trace from previous location where exception was thrown ---
   at asyncexceptions.Program.<RecursiveAsync>d__2.MoveNext() in C:\Work\Things\asyncexceptions\Program.cs:line 32
--- End of stack trace from previous location where exception was thrown ---
   at asyncexceptions.Program.<RecursiveAsync>d__2.MoveNext() in C:\Work\Things\asyncexceptions\Program.cs:line 32
--- End of stack trace from previous location where exception was thrown ---
   at asyncexceptions.Program.<MethodAsync>d__1.MoveNext() in C:\Work\Things\asyncexceptions\Program.cs:line 22
--- End of stack trace from previous location where exception was thrown ---
   at asyncexceptions.Program.<Main>d__0.MoveNext() in C:\Work\Things\asyncexceptions\Program.cs:line 12
```
Post
```
System.Exception: Exception of type 'System.Exception' was thrown.
   at asyncexceptions.Program.<RecursiveAsync>d__2.MoveNext() in C:\Work\Things\asyncexceptions\Program.cs:line 32
   at asyncexceptions.Program.<RecursiveAsync>d__2.MoveNext() in C:\Work\Things\asyncexceptions\Program.cs:line 32
   at asyncexceptions.Program.<RecursiveAsync>d__2.MoveNext() in C:\Work\Things\asyncexceptions\Program.cs:line 32
   at asyncexceptions.Program.<RecursiveAsync>d__2.MoveNext() in C:\Work\Things\asyncexceptions\Program.cs:line 32
   at asyncexceptions.Program.<RecursiveAsync>d__2.MoveNext() in C:\Work\Things\asyncexceptions\Program.cs:line 32
   at asyncexceptions.Program.<RecursiveAsync>d__2.MoveNext() in C:\Work\Things\asyncexceptions\Program.cs:line 32
   at asyncexceptions.Program.<MethodAsync>d__1.MoveNext() in C:\Work\Things\asyncexceptions\Program.cs:line 22
   at asyncexceptions.Program.<Main>d__0.MoveNext() in C:\Work\Things\asyncexceptions\Program.cs:line 12
```

**Comment by benaadams** [comment]
Test
```csharp
class Program
{
    static async Task Main(string[] args)
    {
        try
        {
            await MethodAsync();
        }
        catch (Exception e)
        {
            Console.WriteLine(e);
        }
    }

    static async Task MethodAsync()
    {
        await RecursiveAsync(5);
    }

    static async Task RecursiveAsync(int count)
    {
        if (count == 0)
        {
            throw new Exception();
        }

        await RecursiveAsync(count - 1);
    }
}
```

**Comment by stephentoub** [comment]
Ok, thanks, looks ok to me.

**Comment by benaadams** [comment]
netfx-port-consider?
